### PR TITLE
NormalizeNamespaceByPSR4ComposerAutoloadRector: Replace references correctly

### PR DESCRIPTION
**Pull request by jtojnar**
Previously, `NormalizeNamespaceByPSR4ComposerAutoloadRector` did not update references to other normalized classes properly. This change fixes that by recording the renames using `RenamedClassesDataCollector`.

Supersedes https://github.com/rectorphp/rector-src/pull/2482

### CODE REVIEW

File: e2e/psr-4/rector.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+use Rector\PSR4\Rector\FileWithoutNamespace\NormalizeNamespaceByPSR4ComposerAutoloadRector;
+use Rector\PostRector\Rector\ClassRenamingPostRector;
+
+return static function (Rector\Config\RectorConfig $rectorConfig): void {
+    $rectorConfig->paths([
+        __DIR__ . '/src',
+    ]);
+
+    $rectorConfig->rule(NormalizeNamespaceByPSR4ComposerAutoloadRector::class);
+    $rectorConfig->rule(ClassRenamingPostRector::class);
```
Comment (jtojnar) [inline e2e/psr-4/rector.php]: Not sure if this is needed but it does not work either way – it does not look like the `ClassRenamingPostRector` does anything to fully-qualified name references.

Comment (jtojnar) [reply e2e/psr-4/rector.php]: Apparently, it is not necessary. And it is skipping the processing here:

https://github.com/rectorphp/rector-src/blob/02cec7ebc43469d599cc6a66945e239267626ed6/rules/Renaming/NodeManipulator/ClassRenamer.php#L164-L166

Comment (jtojnar) [reply e2e/psr-4/rector.php]: Apparently, the `$parentNode` is `null`. Possibly `FullyQualifyStmtsAnalyzer` is removing the attribute?

Comment (samsonasik) [reply e2e/psr-4/rector.php]: `FullyQualifyStmtsAnalyzer->process()` produce create new `FullyQualified` instance, the attribute may need to be mirrored:

```php
return new FullyQualified($name, $node->getAttributes());
```

File: rules/PSR4/Rector/FileWithoutNamespace/NormalizeNamespaceByPSR4ComposerAutoloadRector.php [hunk 2]
```diff
@@ -159,4 +167,41 @@ private function refactorFileWithoutNamespace(
...
+    private function findNonAnonymousClassLikes(Node $node): array
+    {
+        $classLikes = $this->betterNodeFinder->findInstanceOf([$node], ClassLike::class);
+
+        return array_filter($classLikes, static function (ClassLike $classLike): bool {
+            if (! $classLike instanceof Class_) {
+                return true;
+            }
+
+            return ! $classLike->isAnonymous();
```
Comment (samsonasik) [inline rules/PSR4/Rector/FileWithoutNamespace/NormalizeNamespaceByPSR4ComposerAutoloadRector.php]: there is `ClassAnalyzer->isAnonymousClass()` for it for more wider use cases.

Comment (jtojnar) [reply rules/PSR4/Rector/FileWithoutNamespace/NormalizeNamespaceByPSR4ComposerAutoloadRector.php]: This is copied from https://github.com/rectorphp/rector-src/blob/f9916a10977a7000c126a01e17bf7c6f69ccf737/rules/PSR4/Rector/Namespace_/MultipleClassFileToPsr4ClassesRector.php#L174-L185, I guess it should be extracted somewhere.

File: e2e/psr-4/rector.php (hunk 1 above)
Comment (jtojnar) [reply e2e/psr-4/rector.php]: Even then, it still does not seem to work properly – only fixes references in one of the files (weird, they are structurally identical). Will need to do more digging.

Comment (TomasVotruba) [reply e2e/psr-4/rector.php]: Possibly, you can re-use `RenamedClassesDataCollector`. Add old -> new names there and it should be picked up in `ClassRenamingPostRector` and resolve this for you.

The post-Rector are run after all other rules, to post-process the final printer file code.

**Comment by TomasVotruba** [comment]
Do you need help with this one? It seems almost finished :slightly_smiling_face:

**Comment by jtojnar** [comment]
I have yet to find time to investigate why it still does not work. Not sure when I will be able to work on this.

File: rules/PSR4/NodeManipulator/FullyQualifyStmtsAnalyzer.php [hunk 3]
```diff
@@ -65,7 +65,7 @@ public function process(array $stmts): void
                 return null;
             }
 
-            return new FullyQualified($name);
+            return new FullyQualified($name, $node->getAttributes());
```
Comment (samsonasik) [inline rules/PSR4/NodeManipulator/FullyQualifyStmtsAnalyzer.php]: better to only map PARENT attribute, as mirror all attributes may cause error on some use cases when checking origNode:

```php
\assert($class === \get_class($origNode));
```

```suggestion
            $fullyQualified = new FullyQualified($name);
            $fullyQualified->setAttribute(Attributekey::PARENT_NODE, $parent);
            
            return $fullyQualified;
```

File: rules/PSR4/Rector/FileWithoutNamespace/NormalizeNamespaceByPSR4ComposerAutoloadRector.php (hunk 2 above)
Comment (jtojnar) [reply rules/PSR4/Rector/FileWithoutNamespace/NormalizeNamespaceByPSR4ComposerAutoloadRector.php]: Extracted to `BetterNodeFinder` and switched to that.

File: rules/PSR4/NodeManipulator/FullyQualifyStmtsAnalyzer.php (hunk 3 above)
Comment (jtojnar) [reply rules/PSR4/NodeManipulator/FullyQualifyStmtsAnalyzer.php]: Done.

File: e2e/psr-4/rector.php (hunk 1 above)
Comment (jtojnar) [reply e2e/psr-4/rector.php]: I am using `RenamedClassesDataCollector` now and it still updates the references only in one of the files correctly but not the other. Possibly by that time the namespace of the referenced file is changed so the rename no longer finds it.

File: rules/PSR4/NodeManipulator/FullyQualifyStmtsAnalyzer.php [hunk 4]
```diff
@@ -65,7 +65,10 @@ public function process(array $stmts): void
                 return null;
             }
 
-            return new FullyQualified($name);
+            $fullyQualified = new FullyQualified($name);
+            $fullyQualified->setAttribute(Attributekey::PARENT_NODE, $parentNode);
+
+            return $fullyQualified;
```
Comment (samsonasik) [inline rules/PSR4/NodeManipulator/FullyQualifyStmtsAnalyzer.php]: this parent attribute mirror is no longer needed, covered at https://github.com/rectorphp/rector-src/pull/2962/files#diff-66beba605a2f40ccb06a352cb8d23db9246969feb7bc25d624cb200f8ba6af12

Comment (jtojnar) [reply rules/PSR4/NodeManipulator/FullyQualifyStmtsAnalyzer.php]: Dropped the commit.

**Comment by TomasVotruba** [comment]
@jtojnar We'd love to have this feature in Rector :)

What is missing here to make CI pass?

**Comment by jtojnar** [comment]
I would expect the code to work as is so there is something wrong with my assumptions.

Unfortunately, I still have not found a time to dive into Rector innards to learn how rules are interleaved, and how `RenamedClassesDataCollector` really interacts with namespaces that are modified.

**Comment by TomasVotruba** [comment]
Closing as the buggy rule is now removed to avoid causing harm.

**Pull request closed without merging**